strawberryfields/program.py

```diff
@@ -458,20 +435,39 @@ def append(self, op, reg):
         self.circuit.append(Command(op, reg))
         return reg
 
+    def _linked_copy(self):
+        """Create a copy of the Program, linked to the original.
+
+        Both the original and the copy are :meth:`locked <lock>`, since they share their RegRefs.
+
+        Returns:
+            Program: a copy of the Program
+        """
+        self.lock()
+        p = copy.copy(self)  # shares RegRefs with the source
+        # link to the original source Program
+        if self.source is None:
+            p.source = self
+        else:
+            p.source = self.source
+        return p
 
-    def compile(self, backend='fock', **kwargs):
-        """Compile the program for the given backend.
+    def compile(self, target, **kwargs):
+        """Compile the program targeting the given circuit template.
 
-        The compilation step validates the program, making sure all the Operations
-        used are accepted by the target backend.
-        Additionally it may decompose certain gates into sequences of simpler gates.
+        Validates the program against the given target, making sure all the Operations
+        used are accepted by the target template.
+        Additionally, depending on the target, the compilation may modify the quantum circuit
+        into an equivalent circuit, e.g. by decomposing certain gates into sequences
+        of simpler gates, or optimizing the gate ordering using commutation rules.
 
-        The compiled program shares its RegRefs with the original, which makes it easier
-        to access the measurement results, but also necessitates the locking of both the
-        compiled program and the original to make sure the RegRef state remains consistent.
+        The returned compiled Program shares its :class:`RegRefs <RegRef>` with the original,
+        which makes it easier to access the measurement results, but also necessitates the
+        :meth:`locking <lock>` of both the compiled program and the original to make sure the
+        RegRef state remains consistent.
 
         Args:
-            backend (str): target backend
+            target (str, DeviceSpecs): short name of the target circuit template, or the template itself
```

Comment:
The template concept in PennyLane is sufficiently different that I'd rather not use it in SF to avoid confusion. The PL templates are parametrized circuits, the DeviceSpecs "circuit templates" in SF can define much more general circuit classes. But CircuitClass is not a good name either, because people would think of Python classes and get confused...

Reply:
I agree with @smite; 'templates', while similar, might be confused with user-facing parametrized circuit templates. At one point we had a long term goal to add a `templates` module to SF (including things such as CVQNN, GBS, etc) so I would prefer to reserve that name.

I like `CircuitClass` but it is confusing. `CircuitGroup`/`CircuitCategory`/`CircuitSpecies`/`CircuitGenus`/`CircuitRank`?

We could even go with `CircuitTarget` for consistency elsewhere.

Reply:
We should spin this comment thread off into a new issue (or add it to an existing relevant issue).

Reply:
Agreed. No need to hold up PR for this